testsuite/forge/src/result.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright © Aptos Foundation
+// Parts of the project are originally copyright © Meta Platforms, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use std::{fmt::{Display, Formatter}, io::{self, Write as _}};
+use termcolor::{Color, ColorChoice, ColorSpec, StandardStream, WriteColor};
+
+#[derive(Debug, Clone)]
+pub enum TestResult {
+    Ok,
+    FailedWithMsg(String),
+}
+
+impl Display for TestResult {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        match self {
+            TestResult::Ok => write!(f, "Test Ok"),
+            TestResult::FailedWithMsg(msg) => write!(f, "Test Failed: {}", msg),
+        }
+    }
+}
+
+pub trait TestObserver {
+    fn handle_result(&self, test_name: String, result: &TestResult) -> Result<()>;
+    fn finish(&self) -> Result<()>;
+}
+
+pub struct TestSummary {
+    stdout: StandardStream,
+    total: usize,
+    filtered_out: usize,
+    passed: usize,
+    failed: Vec<String>,
+    observers: Vec<Box<dyn TestObserver>>,
+}
+
+impl TestSummary {
+    pub fn new(total: usize, filtered_out: usize) -> Self {
+        Self {
+            stdout: StandardStream::stdout(ColorChoice::Auto),
+            total,
+            filtered_out,
+            passed: 0,
+            failed: Vec::new(),
+            observers: Vec::new(),
+        }
+    }
+
+    pub fn add_observer(&mut self, observer: Box<dyn TestObserver>) {
+        self.observers.push(observer);
+    }
+
+    pub fn handle_result(&mut self, name: String, result: TestResult) -> io::Result<()> {
+        write!(self.stdout, "test {} ... ", name)?;
```

Comment:
Fancy fancy. I suppose these tests only run sequentially, otherwise the output might get all mixed up right?

Reply:
Most of this is moved :P so a lot of this is not mine